### ISSUE TITLE
Licensing and other questions

### ISSUE THREAD

**Issue by dheijl** [yotarok/flacenc-rs#2]
Hi again,

I have now forked your code and would like to ask you some questions before I proceed.

- Licensing
    My own code is MIT licensed (swyh-rs), your code is Apache licensed. If I modify your code in my fork I suppose I would have to switch to a dual license. What are your thoughts on this?
- Modifying your code in my fork for my needs
    - I would like to get rid of libsndfile, by hiding it behind a feature. I have no use for it, and it is a dependency I would like to avoid. If I use libsndfile I can as well use libflac-sys or libflac-bound, and the reason I was looking for a pure Rust implementation of Flac encoding was to avoid the use of libflac. What are your views on this? libsndfile is only used for testing AFAICS.
    - I will have to make some structs public (like MetaDataBlock and possibly others) if I want to use the low-level API of flacenc-rs.
    - I will probably have to extend or change or write an alternative for PreloadedSignal, so that is behaves like a buffer for writing and reading samples

All this leads me to the following question: do I plod ahead in my fork on my own or are you interested in integrating my changes in one way or another in your code (me creating pull requests and you commenting and requesting changes before approving etc...)

Looking forward to your opinion on this.

**Comment by yotarok** [comment]
As an owner of this project, I encourage you to contribute to the main branch via PR rather than making a fork. However, this product is also an unofficial Google product, so you will need to submit Google Contributor License Agreement (CLA; https://cla.developers.google.com/).

In case if you are not okay with CLA, I can modify the code on behalf of you for making it more accessible by e.g.:
 1. making libsndfile dependency optional (in library mode)
 2. making some structs public (please let me know which)

With the above changes, I think you can be integrate this library in your project without making a fork.

**Comment by dheijl** [comment]
I have no problem with the CLA, done.

And I totally agree that it is better to contribute with PR's.

I suppose that the only struct that needs to be public for me to use the library is MetaDataBlock.

I had a go at making libsndfile and clap optional using a feature and #[cfg], and it works but it makes the code messy and ugly with all these #[cfg] in my opinion, Rust can apparently not cope gracefully with both a main.rs and a lib.rs with different dependencies in the same crate.

Would you be prepared to split the crate  into a library (core) and a binary crate?

**Comment by yotarok** [comment]
Thanks, I'm fine with splitting the crate.
I prefer to keep the main library logic in the root crate, and to move "main.rs" to, e.g., "cli" sub-crate.

Please prepare two PRs one for splitting, and the other one for the rest.

**Comment by dheijl** [comment]
Thanks! I'll try the "split" PR first before I start anything else.

**Comment by dheijl** [comment]
> Please prepare two PRs one for splitting, and the other one for the rest.

I don't think it can be done in seperate PR's, because in my opinion part of `components.rs` should go in lib (`BitRepr`, `MetaDataBlock `related) while other parts (file and `Stream `related) could (should?) go to bin.  

What is your opinion on this?

**Comment by yotarok** [comment]
I think everything except for "main.rs" can be in the library.
Not in your case, but `Stream` may also have library use cases.
There're only few CLI-specific functions outside of "main.rs", so I think it's fine.
(e.g. serde-dependency in "config.rs" may not be relevant for library usage, but I think for now it's okay.)

I suggested preparing two PRs because I prefer one commit per PR.
But, now I recognize that this case actually requires commit dependency, so please ignore the previous advise.
I now think you should use one PR, with 2 commits.

**Comment by dheijl** [comment]
I agree that file support should be in the library, but because of the dependency on libsndfile I think it should also be an optional feature of the library.
I'll prepare a PR that splits off main.rs but leaves everything else as is, is that OK?
And perhaps we can discuss the libsndfile issue after that?
